src/rest/applications/mcm/resubmission.py

```diff
@@ -0,0 +1,428 @@
+        if not target_campaign:
+            msg = (
+                f"There's no campaign related to the desired data tier ({datatier}) in "
+                f"the linked chained campaign ({member_of_campaign}) that could "
+                "be used to reserve the chained request"
+            )
+            raise ValueError(msg)
+
+        # Reserve the chain
+        reserve_endpoint = f"restapi/chained_requests/flow/{chain_request_prepid}/reserve/{target_campaign}"
+        reserve_result = self._mcm._get(url=reserve_endpoint)
+        if not reserve_result or not reserve_result.get("results", False):
+            raise RuntimeError(
+                f"Unable to reserve the chained request ({chain_request_prepid}). Details: {reserve_result}"
+            )
+
+    def _approve_request_until(
+        self, request_prepid: str, approval: str, status: str
+    ) -> None:
+        """
+        Approves one request until a desired state.
+
+        Args:
+            request_prepid: Request identifier.
+            approval: Desired approval for the request.
+            status: Desired status for the request.
+        """
+        request: Union[dict, None] = self._mcm.get(
+            object_type="requests", object_id=request_prepid
+        )
+        if not request:
+            raise ValueError(f"Request not found: {request_prepid}")
+
+        is_root: bool = self._mcm.is_root_request(request=request)
+        attempts = 10
+        for _ in range(attempts):
+            request = self._mcm.get(object_type="requests", object_id=request_prepid)
+            req_approval = request.get("approval")
+            req_status = request.get("status")
+
+            if approval == req_approval and status == req_status:
+                return
+
+            approve_result = self._mcm.approve(
+                object_type="requests", object_id=request_prepid
+            )
+            if not approve_result or not approve_result.get("results"):
+                if approve_result and is_root:
+                    message = approve_result.get("message", "")
+                    if "Illegal Approval Step: 5" in message:
+                        return
+
+                msg = (
+                    "Unable to approve the request to the next status - "
+                    f"Request PrepID: {request_prepid} - "
+                    f"Current approval/status: {req_approval}/{req_status} - "
+                    f"Details: {pformat(approve_result)}"
+                )
+                self.logger.error(msg)
+                raise RuntimeError(msg)
+
+        raise RuntimeError("Unable to get the desired approval status")
+
+    def _perform_chain_request_injection(
+        self, chain_request_prepid: str, root_request_processing: RootRequestReset
+    ) -> None:
+        """
+        Injects a chained request based on how its root request was reset.
+        This MUST BE executed only after rewinding a chained request to root and reserving it again.
```

Comment:
This is a bit scary... what happens if there is an exception in the middle of the script? Does all state remain self-consistent?